rustworkx-core/src/dag_algo.rs

```diff
@@ -313,6 +316,119 @@ where
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected, // Used for .edged_directed
+    <G as GraphBase>::NodeId: Eq + Hash,
+{
+    let mut output_indices: Vec<Vec<G::NodeId>> = Vec::new();
+    let mut cur_layer = first_layer;
+    let mut next_layer: Vec<G::NodeId> = Vec::new();
+    let mut predecessor_count: HashMap<G::NodeId, usize> = HashMap::new();
+    let node_set = graph.node_identifiers().collect::<HashSet<G::NodeId>>();
+    // Throw error if a cycle is found at the current node
+    if let Some(node) = &cur_layer.first() {
+        let check_cycle = find_cycle(&graph, Some(**node));
+        if !check_cycle.is_empty() {
+            return Err(LayersError(Some(
+                "The provided graph has a cycle".to_string(),
+            )));
+        }
+    }
+    for layer_node in &cur_layer {
+        if !node_set.contains(layer_node) {
+            return Err(LayersError(Some(format!(
+                "An index input in 'first_layer' {} is not a valid node index in the graph",
+                graph.to_index(*layer_node)
+            ))));
+        }
+    }
+    output_indices.push(cur_layer.to_owned());
+
+    // Iterate until there are no more
+    while !cur_layer.is_empty() {
+        for node in &cur_layer {
+            let children = graph.neighbors_directed(*node, petgraph::Direction::Outgoing);
+            let mut used_indices: HashSet<G::NodeId> = HashSet::new();
+            for succ in children {
+                // Skip duplicate successors
+                if used_indices.contains(&succ) {
+                    continue;
+                }
+                used_indices.insert(succ);
+                let mut multiplicity: usize = 0;
+                let raw_edges: G::EdgesDirected =
+                    graph.edges_directed(*node, petgraph::Direction::Outgoing);
+                for edge in raw_edges {
+                    if edge.target() == succ {
+                        multiplicity += 1;
+                    }
+                }
+                predecessor_count
+                    .entry(succ)
+                    .and_modify(|e| *e -= multiplicity)
+                    .or_insert(
+                        // Get the number of incoming edges to the successor
+                        graph
+                            .edges_directed(succ, petgraph::Direction::Incoming)
+                            .count()
+                            - multiplicity,
+                    );
+                if *predecessor_count.get(&succ).unwrap() == 0 {
+                    next_layer.push(succ);
+                    predecessor_count.remove(&succ);
+                }
+            }
+        }
+        if !next_layer.is_empty() {
+            output_indices.push(next_layer.to_owned());
+        }
+        cur_layer = next_layer;
+        next_layer = Vec::new();
+    }
+    Ok(output_indices)
+}
```

Comment:
The iterator comment I referred to in the original issue can actually be done here. It requires a bit of a restructuring of the code, you have to move the tracking variables into a struct and implement the `Iterator` trait for that struct. This ends up looking something like:

```suggestion
    Ok(LayersIterator {
        graph,
        cur_layer,
        output_indices,
        next_layer,
        predecessor_count,
        node_set,
        first_iter: false,
    })
}

struct LayersIterator<G, N> {
    graph: G,
    cur_layer: Vec<N>,
    output_indices: Vec<Vec<N>>,
    next_layer: Vec<N>,
    predecessor_count: HashMap<N, usize>,
    node_set: HashSet<N>,
    first_iter: bool,
}

impl<G, N> Iterator for LayersIterator<G, N>
where
    G: NodeIndexable // Used in from_index and to_index.
        + NodeCount // Used in find_cycle
        + EdgeCount // Used in find_cycle
        + Visitable // Used in find_cycle
        + IntoNodeIdentifiers<NodeId = N> // Used for .node_identifiers
        + IntoNeighborsDirected // Used for .neighbors_directed
        + IntoEdgesDirected // Used for .edged_directed
        + GraphBase<NodeId = N>,
    N: Copy + Eq + Hash,
{
    type Item = Vec<N>;
    fn next(&mut self) -> Option<Self::Item> {
        if !self.first_iter {
            self.first_iter = true;
            Some(self.cur_layer.clone())
        } else if self.cur_layer.is_empty() {
            None
        } else {
            while !self.cur_layer.is_empty() {
                for node in &self.cur_layer {
                    let children = self
                        .graph
                        .neighbors_directed(*node, petgraph::Direction::Outgoing);
                    let mut used_indices: HashSet<N> = HashSet::new();
                    for succ in children {
                        // Skip duplicate successors
                        if used_indices.contains(&succ) {
                            continue;
                        }
                        used_indices.insert(succ);
                        let mut multiplicity: usize = 0;
                        let raw_edges: G::EdgesDirected = self
                            .graph
                            .edges_directed(*node, petgraph::Direction::Outgoing);
                        for edge in raw_edges {
                            if edge.target() == succ {
                                multiplicity += 1;
                            }
                        }
                        self.predecessor_count
                            .entry(succ)
                            .and_modify(|e| *e -= multiplicity)
                            .or_insert(
                                // Get the number of incoming edges to the successor
                                self.graph
                                    .edges_directed(succ, petgraph::Direction::Incoming)
                                    .count()
                                    - multiplicity,
                            );
                        if *self.predecessor_count.get(&succ).unwrap() == 0 {
                            self.next_layer.push(succ);
                            self.predecessor_count.remove(&succ);
                        }
                    }
                }
                if !self.next_layer.is_empty() {
                    return Some(self.next_layer.to_owned());
                }
                std::mem::swap(&mut self.cur_layer, &mut self.next_layer);
                self.next_layer.clear();
            }
            None
        }
    }
}
```

Then you return the return type to 

```rust
impl Iterator<Item=Vec<G::NodeId>>
```

I think having the true iterator here would make sense because we can stop early if someone hits the target, or we can operate on the data in place as we iterate which I think would be more powerful than collecting into a Vec.

Note, I haven't tested the functionality of this suggestion, just that it compiles locally. But there are some warnings so I probably messed up something in the migration. This is more the basic structure to shoot for.

Reply:
This makes sense, I'll have something like this ready soon!